File: project/drivers/driver/r/R/scram.R
# Copyright (c) 2026 ScratchBird Software Inc.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# SPDX-License-Identifier: MPL-2.0

sb_scram_client <- function(username, algorithm = "sha256") {
  nonce <- openssl::base64_encode(openssl::rand_bytes(18))
  list(
    username = username,
    client_nonce = nonce,
    client_first_bare = "",
    server_signature = NULL,
    algorithm = algorithm
  )
}

raw_xor <- function(a, b) {
  as.raw(bitwXor(as.integer(a), as.integer(b)))
}

hmac_hash <- function(algorithm, key, message) {
  key_raw <- if (is.raw(key)) key else charToRaw(as.character(key))
  msg_raw <- if (is.raw(message)) message else charToRaw(as.character(message))
  if (identical(algorithm, "sha512")) {
    hash_fn <- openssl::sha512
    block_size <- 128L
  } else {
    hash_fn <- openssl::sha256
    block_size <- 64L
  }
  if (length(key_raw) > block_size) {
    key_raw <- hash_fn(key_raw)
  }
  if (length(key_raw) < block_size) {
    key_raw <- c(key_raw, as.raw(rep(0L, block_size - length(key_raw))))
  }
  o_key <- raw_xor(key_raw, as.raw(rep(0x5c, block_size)))
  i_key <- raw_xor(key_raw, as.raw(rep(0x36, block_size)))
  inner <- hash_fn(c(i_key, msg_raw))
  hash_fn(c(o_key, inner))
}

int_to_raw_be <- function(value) {
  con <- rawConnection(raw(), "wb")
  on.exit(close(con))
  writeBin(as.integer(value), con, size = 4, endian = "big")
  rawConnectionValue(con)
}

pbkdf2_hmac_hash <- function(password, salt, iterations, keylen, algorithm = "sha256") {
  hlen <- if (identical(algorithm, "sha512")) 64L else 32L
  blocks <- ceiling(keylen / hlen)
  out <- raw()
  for (i in seq_len(blocks)) {
    u <- hmac_hash(algorithm, password, c(salt, int_to_raw_be(i)))
    t <- u
    if (iterations > 1) {
      for (j in 2:iterations) {
        u <- hmac_hash(algorithm, password, u)
        t <- raw_xor(t, u)
      }
    }
    out <- c(out, t)
  }
  out[seq_len(keylen)]
}

sb_scram_client_first <- function(state) {
  escaped <- gsub("=", "=3D", gsub(",", "=2C", state$username, fixed = TRUE), fixed = TRUE)
  state$client_first_bare <- paste0("n=", escaped, ",r=", state$client_nonce)
  list(state = state, message = paste0("n,,", state$client_first_bare))
}

sb_scram_handle_server_first <- function(state, password, server_first) {
  attrs <- parse_scram_attrs(server_first)
  nonce <- attrs$r
  if (is.null(nonce) || !startsWith(nonce, state$client_nonce)) stop("SCRAM server nonce mismatch")
  salt_b64 <- attrs$s
  iter_str <- attrs$i
  if (is.null(salt_b64) || is.null(iter_str)) stop("SCRAM server-first missing fields")
  iterations <- as.integer(iter_str)
  salt <- openssl::base64_decode(salt_b64)
  algorithm <- if (!is.null(state$algorithm)) state$algorithm else "sha256"
  keylen <- if (identical(algorithm, "sha512")) 64L else 32L
  hash_fn <- if (identical(algorithm, "sha512")) openssl::sha512 else openssl::sha256
  salted <- pbkdf2_hmac_hash(password, salt, iterations, keylen = keylen, algorithm = algorithm)
  client_key <- hmac_hash(algorithm, salted, "Client Key")
  stored_key <- hash_fn(client_key)
  client_final_without_proof <- paste0("c=biws,r=", nonce)
  auth_message <- paste(state$client_first_bare, server_first, client_final_without_proof, sep = ",")
  client_signature <- hmac_hash(algorithm, stored_key, auth_message)
  client_proof <- raw_xor(client_key, client_signature)
  server_key <- hmac_hash(algorithm, salted, "Server Key")
  state$server_signature <- hmac_hash(algorithm, server_key, auth_message)
  proof_b64 <- openssl::base64_encode(as.raw(client_proof))
  list(state = state, message = paste0(client_final_without_proof, ",p=", proof_b64))
}

sb_scram_verify_server_final <- function(state, server_final) {
  attrs <- parse_scram_attrs(server_final)
  verifier <- attrs$v
  if (is.null(verifier)) stop("SCRAM server-final missing verifier")
  expected <- openssl::base64_encode(state$server_signature)
  if (verifier != expected) stop("SCRAM server signature mismatch")
  TRUE
}

parse_scram_attrs <- function(message) {
  attrs <- list()
  parts <- strsplit(message, ",", fixed = TRUE)[[1]]
  for (part in parts) {
    if (part == "") next
    kv <- strsplit(part, "=", fixed = TRUE)[[1]]
    if (length(kv) < 2) next
    attrs[[kv[1]]] <- kv[2]
  }
  attrs
}

File: project/drivers/driver/dart/lib/src/scram.dart
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

import 'errors.dart';

enum ScramAlgorithm {
  sha256,
  sha512,
}

class ScramClient {
  final String username;
  final ScramAlgorithm algorithm;
  late String clientNonce;
  String? clientFirstBare;
  Uint8List? serverSignature;

  ScramClient(this.username, {this.algorithm = ScramAlgorithm.sha256}) {
    final rand = Random.secure();
    final bytes = List<int>.generate(18, (_) => rand.nextInt(256));
    clientNonce = base64.encode(bytes);
  }

  String clientFirstMessage() {
    clientFirstBare = 'n=${_escape(username)},r=$clientNonce';
    return 'n,,' + clientFirstBare!;
  }

  String handleServerFirst(String password, String serverFirst) {
    final attrs = _parseAttrs(serverFirst);
    final nonce = attrs['r'] ?? '';
    final saltB64 = attrs['s'] ?? '';
    final iterStr = attrs['i'] ?? '0';
    if (!nonce.startsWith(clientNonce)) {
      throw const ScratchBirdAuthException('SCRAM server nonce mismatch');
    }
    final iterations = int.parse(iterStr);
    final salt = base64.decode(saltB64);
    final keyLen = algorithm == ScramAlgorithm.sha512 ? 64 : 32;
    final salted = _pbkdf2(utf8.encode(password), salt, iterations, keyLen);
    final clientKey = _hmac(salted, utf8.encode('Client Key'));
    final storedKey = _hash(clientKey);
    final clientFinalWithoutProof = 'c=biws,r=$nonce';
    final authMessage =
        '${clientFirstBare!},$serverFirst,$clientFinalWithoutProof';
    final clientSignature = _hmac(storedKey, utf8.encode(authMessage));
    final clientProof = _xor(clientKey, clientSignature);
    final serverKey = _hmac(salted, utf8.encode('Server Key'));
    serverSignature = _hmac(serverKey, utf8.encode(authMessage));
    return '$clientFinalWithoutProof,p=${base64.encode(clientProof)}';
  }

  void verifyServerFinal(String serverFinal) {
    final attrs = _parseAttrs(serverFinal);
    final verifier = attrs['v'] ?? '';
    final expected = base64.encode(serverSignature ?? Uint8List(0));
    if (verifier != expected) {
      throw const ScratchBirdAuthException('SCRAM server signature mismatch');
    }
  }

  String _escape(String input) =>
      input.replaceAll('=', '=3D').replaceAll(',', '=2C');

  Map<String, String> _parseAttrs(String message) {
    final out = <String, String>{};
    for (final part in message.split(',')) {
      final idx = part.indexOf('=');
      if (idx > 0) {
        out[part.substring(0, idx)] = part.substring(idx + 1);
      }
    }
    return out;
  }

  Uint8List _hmac(List<int> key, List<int> data) {
    final hmacSha = Hmac(
      algorithm == ScramAlgorithm.sha512 ? sha512 : sha256,
      key,
    );
    return Uint8List.fromList(hmacSha.convert(data).bytes);
  }

  Uint8List _hash(List<int> data) {
    final digest = algorithm == ScramAlgorithm.sha512 ? sha512 : sha256;
    return Uint8List.fromList(digest.convert(data).bytes);
  }

  Uint8List _xor(List<int> left, List<int> right) {
    final out = Uint8List(left.length);
    for (var i = 0; i < left.length; i++) {
      out[i] = left[i] ^ right[i];
    }
    return out;
  }

  Uint8List _pbkdf2(
    List<int> password,
    List<int> salt,
    int iterations,
    int keyLen,
  ) {
    final blockSize = algorithm == ScramAlgorithm.sha512 ? 64 : 32;
    final blocks = (keyLen / blockSize).ceil();
    final out = BytesBuilder();
    for (var i = 1; i <= blocks; i++) {
      out.add(_pbkdf2F(password, salt, iterations, i));
    }
    final bytes = out.toBytes();
    return Uint8List.sublistView(bytes, 0, keyLen);
  }

  List<int> _pbkdf2F(
      List<int> password, List<int> salt, int iterations, int blockIndex) {
    final block = Uint8List(salt.length + 4);
    block.setAll(0, salt);
    block[block.length - 4] = (blockIndex >> 24) & 0xff;
    block[block.length - 3] = (blockIndex >> 16) & 0xff;
    block[block.length - 2] = (blockIndex >> 8) & 0xff;
    block[block.length - 1] = blockIndex & 0xff;
    var u = _hmac(password, block);
    final out = List<int>.from(u);
    for (var i = 1; i < iterations; i++) {
      u = _hmac(password, u);
      for (var j = 0; j < out.length; j++) {
        out[j] ^= u[j];
      }
    }
    return out;
  }
}
